### DISCUSSION TITLE
How to change RPU config?

### DISCUSSION THREAD

**Discussion by maicoldubbio** [IBM/aihwkit#571] [Q&A] [answered]
Hi there!

Suppose you've trained a model in a hardware-aware fashion, and saved it. As such, when you `load_state_dict`, RPU config will be loaded as well since `load_rpu_config=True` by default.

Now, let's say I want to change some parameters of the RPU config such as `noise_model`: how can you do that? Maybe if there's a way to get the current RPU, change it, and reload it into the tile.

So far, I've been using `conver_to_analog` after loading the analog model, specifying a new RPU config. However I guess this is incorrect, because things like digital scales coming from HWA and other parameters will be lost.

How can I do that? What is the most correct way?

**Comment by maljoras** [comment, marked as answer]
Hi @edminge,
thanks for the question. The most general way is, if you have a checkpoint file, to construct a new model with a new `RPUConfig` and do 
```python 
analog_model.load_state_dict(state_dict, load_rpu_config=False)
```
However, there is also a convenient function to change the `RPUConfig` of an existing model via: 
```python
for tile in analog_model.analog_tiles():
     tile.to(rpu_config=rpu_config)
```   

This loops through all  the tiles (aka crossbars) and applies the new `RPUConfig`. Note that different tiles could in principle have different `RPUConfigs` to specify e.g. different noise levels etc, so the `RPUConfig` is a tile property and not a model property.

**Reply by maicoldubbio** [reply]
Thanks for the answer @maljoras! And in the 2 ways, will the learnable parameters (scales and so on) be kept? I mean, I wouldn't want to change RPU and spoil the benefits of HWA training. 

IDK whether they're found in the `state_dict` (so kept when changing RPU) or in the RPU itself and so mess up something.

**Reply by maljoras** [reply]
Actually, if you only want to specify the noise model, note that when you call `program_weights` to apply the noise model, a new noise model can be given directly as well: 
```python
analog_model.program_analog_weights(noise_model=noise_model)
```
This would supersede the noise model defined in the `RPUconfig` (and actually would update it to the new model internally)

**Reply by maljoras** [reply]
The learnable parameters will be the same and not changed. The learnable parameters are defined as part of the model and not in the `RPUConfig`

**Reply by maicoldubbio** [reply]
Wonderful! Thanks a lot.

**Reply by HeatPhoenix** [reply]
I can't seem to get the following code to function as mentioned by your post.
```py
for tile in analog_model.analog_tiles():
     tile.to(rpu_config)
```
Doing this on my model simply returns
```
TypeError: Device() received an invalid combination of arguments - got (InferenceRPUConfig), but expected one of:
 * (torch.device device)
      didn't match because some of the arguments have invalid types: (InferenceRPUConfig)
 * (str type, int index)
 ```

At its most basic, I have a simple model with a SingleRPUConfig, to which I want to apply an InferenceRPUConfig (and noise model) post-training, pre-inference. Any idea of what's going wrong here? Has this functionality been changed since this answer was posted?

If necessary, I can provide more code to give a minimum example.

**Reply by maljoras** [reply]
Right, it should be
```python
tile.to(rpu_config=rpu_config)
```

**Reply by HeatPhoenix** [reply]
*First off, thank you for super swift response (and great work on the project, which is very important work).*

After the change, I still receive the following error:
```
TypeError: to() received an invalid combination of arguments - got (rpu_config=InferenceRPUConfig, ), but expected one of:
 * (torch.device device, torch.dtype dtype, bool non_blocking, bool copy, *, torch.memory_format memory_format)
 * (torch.dtype dtype, bool non_blocking, bool copy, *, torch.memory_format memory_format)
 * (Tensor tensor, bool non_blocking, bool copy, *, torch.memory_format memory_format)
 ```
 
 Where `model` is a simple CNN for MNIST (that works in training and evaluation), `rpu_config` is simply `InferenceRPUConfig(noise_model=ReRamWan2022NoiseModel())`.
 
 Secondly, the other method `model.program_analog_weights(noise_model=ReRamWan2022NoiseModel)`, also does not seem to work for me, with the error given being: `TypeError: AnalogLayerBase.program_analog_weights() got an unexpected keyword argument 'noise_model'`

**Reply by maljoras** [reply]
From first glance it looks like you are not using the latest version. Make sure you are using the latest version.

**Reply by HeatPhoenix** [reply]
```
# packages in environment at /home/heatphoenix/miniforge3/envs/aihwkit:
aihwkit-gpu               0.8.0           py311hbd5a262_0    conda-forge
```
releases also marks 0.8.0 as the most recent, so I think I'm up to date. Working on linux, using conda. Installed late last week.

**Reply by maljoras** [reply]
I see, likely you need the newest master branch for this. There will be a new release soon on conda, right @kaoutar55?

**Reply by HeatPhoenix** [reply]
I see it here: https://github.com/IBM/aihwkit/blob/8b75f7694e1764912be423ce5635c74f3c8e0a46/src/aihwkit/simulator/tiles/module.py#L204

Thanks for the assistance, I'll just install latest from master branch locally and see how it goes.

**Reply by HeatPhoenix** [reply]
Follow-up question (after installing 0.9.0 from git), perhaps off-topic, perhaps not:

`aihwkit.exceptions.TileModuleError: RPU config mismatch: Cannot replace InferenceRPUConfig with ReRamESPreset`
What are the limitations to replacing RPU configs in this way? Can I replace a SingleRPUConfig with InferenceRPUConfig, if not, why not? 

(as in, I want to train with ReRAM config, to for example have corrupted devices during training, but for evaluation I want realistic device noise models to be applied).

Similarly, doing `model.program_analog_weights(noise_model=ReRamWan2022NoiseModel)`, in this situation will get `TileWithPeriphery.program_weights() got an unexpected keyword argument 'noise_model'`, even when in evaluation mode.

In what way could I still get to the desired result of switching RPUConfigs (train with a ReRAM preset that lets me set corrupted devices, and inference/evaluation with an InferenceRPUConfig/noise model for realistic noise models).

Many thanks, again.

**Reply by maljoras** [reply]
`InferenceRPUConfig` and `ReRamESPreset` are fundamentally different concepts. The first is for chips that are used for inference only (including using ReRAM), which means you program a software model once and use crossbars only for inference (only performing the forward pass in-memory). The latter is a *training* preset, that means it actually performs the full training process in-memory (forward / backward / update) assuming NVMs. That is likely not what you want. You should always use `InferenceRPUConfig` but change the noise_model to e.g. the ReRam noise model (see e.g. `aihwkit.inference.noise.reram`). You could add corrupt devices there. If you want to train with corrupt devices, you can look at the `modifiers` that in some cases support corrupt device training (see https://github.com/IBM/aihwkit/blob/master/src/aihwkit/simulator/parameters/inference.py#L136). 


Doing training with the `aihwkit` while using the `InferenceRPUConfig` means that you perform hardware-aware training (thus preparing the model in software for deployment on the inference-only chip). It thus assumes a perfect update / backward pass.

**Reply by HeatPhoenix** [reply]
I believe using InferenceRPUConfig, with corrupted devices and *then* training (as in, training but taking the corrupted devices into account while training) is the desired result. I'll give that a go using the modifiers. Thanks again. 

Out of curiosity, why is InferenceRPUConfig and ReamESPreset (for example, or any other RPUConfig) split like this? Wouldn't you want (to have the option of) a realistic forward pass with this noise when doing this training on the devices themselves? Does this imply the forward pass in the ReRamESPreset both during training and during evaluation is not _as_ realistic as when using InferenceRPUConfig? 

I understand that InferenceRPUConfig, for example, does not include the ability to set specialized material settings and AIMC SGD algorithms for training, but that would just make the non-Inference only configs supersets of the Inference only configs, no?

**Comment by HeatPhoenix** [comment]
> `InferenceRPUConfig` and `ReRamESPreset` are fundamentally different concepts. The first is for chips that are used for inference only (including using ReRAM), which means you program a software model once and use crossbars only for inference (only performing the forward pass in-memory). The latter is a _training_ preset, that means it actually performs the full training process in-memory (forward / backward / update) assuming NVMs. That is likely not what you want. You should always use `InferenceRPUConfig` but change the noise_model to e.g. the ReRam noise model (see e.g. `aihwkit.inference.noise.reram`). You could add corrupt devices there. If you want to train with corrupt devices, you can look at the `modifiers` that in some cases support corrupt device training (see https://github.com/IBM/aihwkit/blob/master/src/aihwkit/simulator/parameters/inference.py#L136).
> 
> Doing training with the `aihwkit` while using the `InferenceRPUConfig` means that you perform hardware-aware training (thus preparing the model in software for deployment on the inference-only chip). It thus assumes a perfect update / backward pass.

Apologies the double post, to properly make use of your suggestion, I need to install 0.9.0 _with CUDA-support_, but installing from a local package (aka pulling from git -> `pip install .` or something similar) will only install the CPU version, which will toss an error when trying to use this functionality: `RuntimeError: Error in weight_modifier.cpp:204  Requested WeightModifierType not implemented in CPU version.` 

Any tips for installing this version smoothly?

Secondly, perhaps I am misunderstanding `WeightModifierType`, but is it not possible to apply multiple types and multiple parameters? ex. Discretization **and** PCM_NOISE?

Many thanks.
